src/swish/erlang.ss

```diff
@@ -765,9 +770,70 @@
      [(eq? time 'infinity) ($receive matcher src #f #f)]
      [else (throw `#(timeout-value ,time ,src))]))
 
+  ;; Procedures in io.ss may disable interrupts and call wait-for-io. The yield
+  ;; in wait-for-io then bumps the disable count from 1 to 2. When we resume
+  ;; after complete-io, the code in @yield restores the saved disable count 2
+  ;; and calls @handle-trap-exit, which does nothing since we must wait to call
+  ;; the handler with interrupts enabled. An I/O bound process might never use
+  ;; its quantum or call receive, so we avoid starving the trap-exit handler by
+  ;; checking again just before an I/O procedure re-enables interrupts.
+  (define-syntax with-interrupts-disabled-for-I/O
+    (syntax-rules ()
+      [(_ e1 e2 ...)
+       (dynamic-wind
+         #%disable-interrupts
+         (lambda () e1 e2 ...)
+         (lambda ()
+           (@check-and-enable-interrupts)))]))
+
+  (define (current-disable-count) (#3%$tc-field 'disable-count (#3%$tc)))
+
+  (define (@handle-trap-exit)
+    (when (pcb-flag-test self [exit? #t] [handling-exit? #f])
+      (when (#3%fx= (current-disable-count) 1)
+        (@handle-trap-exit-help))))
+
+  (define (@handle-trap-exit-help)
+    (let ([handle-exit (process-trap-exit)])
+      (cond
+       [(and (procedure? handle-exit) (@get-EXIT-msg)) =>
+        (lambda (msg)
+          (define prior-src (pcb-src self))
+          (define armed? #f)
+          (define content (msg-contents msg))
+          (@remove-msg msg)
+          (enable-interrupts)
```

Comment:
Hmm, I think we rely on record field accessors being inlined elsewhere in erlang.ss. That could be a problem when we compile with cp0 disabled for profiling.

Maybe we should wrap the calls to `pcb` setters in a `no-interrupts`. The code here isn't performance-critical, so there's no harm, but I think the other places were more frequently executed.

FWIW, here's what the `dynamic-wind` looks like after `expand/optimize`. It looks like the inline code for `#2%fxlogbit0` won't take an interrupt either, though it can call `$oops`.

``` scheme
(#2%dynamic-wind
  (lambda ()
    (if armed?.1
        (#2%errorf #f "attempt to reenter trap-exit handler ~s" handle-exit.2)
        (#2%void))
    (set! armed?.1 #t)
    (let ([g1.3 (#3%$top-level-value 'self.lgnnu3lheosakvgyylzmvq5uw-0)])
      (if (#3%$sealed-record? g1.3 '#<record type pcb>)
          (#2%void)
          (#3%$record-oops 'pcb-src-set! g1.3 '#<record type pcb>))
      (#3%$object-set! 'scheme-object g1.3 129 #f))
    (let ([p.4 (#3%$top-level-value 'self.lgnnu3lheosakvgyylzmvq5uw-0)])
      (let ([g0.5 (#2%fxlogbit1 5
                   (begin
                     (if (#3%$sealed-record? p.4 '#<record type pcb>)
                         (#2%void)
                         (#3%$record-oops 'pcb-flags p.4 '#<record type pcb>))
                     (#3%$object-ref 'scheme-object p.4 105)))])
        (if (#3%$sealed-record? p.4 '#<record type pcb>)
            (#2%void)
            (#3%$record-oops 'pcb-flags-set! p.4 '#<record type pcb>))
        (#3%$object-set! 'scheme-object p.4 105 g0.5))))
  (lambda () (handle-exit.2 content))
  (lambda ()
    (let ([g1.6 (#3%$top-level-value 'self.lgnnu3lheosakvgyylzmvq5uw-0)]
          [g2.7 prior-src.0])
      (if (#3%$sealed-record? g1.6 '#<record type pcb>)
          (#2%void)
          (#3%$record-oops 'pcb-src-set! g1.6 '#<record type pcb>))
      (#3%$object-set! 'scheme-object g1.6 129 g2.7))
    (let ([p.8 (#3%$top-level-value 'self.lgnnu3lheosakvgyylzmvq5uw-0)])
      (let ([g0.9 (#2%fxlogbit0 5
                   (begin
                     (if (#3%$sealed-record? p.8 '#<record type pcb>)
                         (#2%void)
                         (#3%$record-oops 'pcb-flags p.8 '#<record type pcb>))
                     (#3%$object-ref 'scheme-object p.8 105)))])
        (if (#3%$sealed-record? p.8 '#<record type pcb>)
            (#2%void)
            (#3%$record-oops 'pcb-flags-set! p.8 '#<record type pcb>))
        (#3%$object-set! 'scheme-object p.8 105 g0.9)))))
```

Reply:
Thanks for the expansion. I'm satisfied with the current behavior.